### ISSUE TITLE
Airflow integration start and complete events don't reflect actual task states

### ISSUE THREAD

**Issue by collado-mike** [OpenLineage/OpenLineage#313]
When the Airflow DAG creates a new DagRun, it immediately posts a new `START` event for each task in the DAG. Once the entire DAG has completed, the `handle_callback` method posts a `COMPLETE` or `FAILED` event for each task, depending on the actual end state of each. An example from a simple DAG looks like this (most fields omitted)

```json
{"job":"etl_orders.if_not_exists","state":"START","inputs":[],"outputs":[]}
{"job":"etl_orders.insert","state":"START","inputs":[],"outputs":[]}
{"job":"etl_orders.if_not_exists","state":"START","inputs":[],"outputs":[]}
{"job":"etl_orders.insert","state":"START","inputs":[],"outputs":[]}
{"job":"etl_orders.insert","state":"COMPLETE","inputs":[{"name":"marquez.food_delivery.tmp_orders","namespace":"bigquery"}],"outputs":[{"name":"marquez.food_delivery.orders","namespace":"bigquery"}]}
{"job":"etl_orders.if_not_exists","state":"COMPLETE","inputs":[],"outputs":[{"name":"marquez.food_delivery.orders","namespace":"bigquery"}]}
{"job":"etl_orders.insert","state":"COMPLETE","inputs":[{"name":"marquez.food_delivery.tmp_orders","namespace":"bigquery"}],"outputs":[{"name":"marquez.food_delivery.orders","namespace":"bigquery"}]}
{"job":"etl_orders.if_not_exists","state":"COMPLETE","inputs":[],"outputs":[{"name":"marquez.food_delivery.orders","namespace":"bigquery"}]}
```

The result of this is a correct and complete lineage graph, but the runs show incorrect dataset versions consumed. For example, if the `etl_orders.insert` job creates version X of the `marquez.food_delivery.orders` dataset and I add a GreatExpectationsOperator to the DAG, I expect the `DataQuality` facet to be appended to version X of the dataset. Instead, since the `OpenLineageValidationAction` posts a `COMPLETE` event immediately, without waiting for the rest of the DAG to complete, the Marquez server sees that the validation job completed before the `etl_orders.insert` job, therefore it assumes that the version of the dataset validated is the version prior (version W).

While the Marquez server could use some heuristics to determine that write job is still running, therefore that the dataset is actively being modified and, thus, that the version of the data being read is the new version, this won't always be true. For instance, a data store that supports transaction isolation can support one job reading the old data while a write job modifies the data concurrently.

The Airflow integration should be modified to post `START` and `COMPLETE` events that reflect the actual start and end times of the tasks. In addition to fixing the dataset version issue, it will also reflect the actual runtimes of the tasks, helping consumers understand when jobs are taking longer than normal or when they're starting later than expected.

- [ ] Maciej to write up proposal

**Comment by mobuchowski** [comment]
I've tried to find way to get notified of task completion in 1.10 via DAG - without success. `LineageBackend` based approach works though. Maybe combination of both will work? The `START` and `FAIL` events would be published by DAG as it's now, and complete ones by `LineageBackend`.

Anyway,

>While the Marquez server could use some heuristics to determine that write job is still running, therefore that the dataset is actively being modified and, thus, that the version of the data being read is the new version, this won't always be true. For instance, a data store that supports transaction isolation can support one job reading the old data while a write job modifies the data concurrently.

The fact that times for `COMPLETE` events have correct time doesn't help you, right? If you depend on processing instead of event time, I can see a lot of problems in the future with race conditions. 

Overengineering out loud: maybe Marquez could "reconciliate" such events - if we get `COMPLETE` event for dataset modification in the past, all the datasets that depend on it and have start time after the `COMPLETE` event finished should point at the new dataset? That would solve the transaction isolation, provided that database supports snapshot isolation.

**Comment by mobuchowski** [comment]
Approach with using both `DAG` and `LineageBackend` turns out to be feasible, so let's go with it. 

On the other hand, Marquez's handling of those type of events remains problematic - we can't always control when we emit events. For example, `dbt` integration always emits both start and complete events when it finishes. Even if we could, depending on order of HTTP requests instead of event time will remain a problem in case when we have multiple independent systems emitting OpenLineage events that work on single dataset. 


Further, even if we could have perfect serializability of events, the semantics of which dataset versions is used isn't always right.
For example:

![](https://i.imgur.com/i9xtDx2.jpg)

If storage system in that case would be `bigquery`, then reader would read version before writer's modification. In contrary, when `storage` would be just raw parquet files on object storage, reader would have inconsistent view. Hovewer, do we want to assume anything here?

**Comment by collado-mike** [comment]
> Hovewer, do we want to assume anything here?

No, that's why I don't want to rely on server-side heuristics. I wonder if there's something in the model we can augment to make the relationships more explicit. E.g., if we know that `TaskB` is dependent on `TaskA` and then that `TaskA` writes `DatasetX` and `taskB` reads `DatasetX`, we can know that `TaskB` relies on the version of `DatasetX` that was written by `TaskA`, not the version prior.

This only works for workflows that have explicit dependencies (e.g., Airflow DAGs) and makes it tricky to support external jobs that work on shared datasets- e.g., an Airflow DAG that modifies a Dataset, then triggers a Spark job that reads the Dataset and writes a new Dataset that is then consumed by another task in the Airflow DAG. 

The parent run hierarchy can help- e.g., the Spark job's parent run will be the Airflow task that triggered it, thus it will inherit the dependency on the task that writes its input Dataset. However, the Spark run will start and complete before its dependency Airflow task, so what version of the Dataset will be recorded for the Spark job? I suppose if we rely on the dependency mechanism, it will assume that the correct version of the dataset was read. Currently, we always assume the version is the `Dataset.currentVersion`; we can change that assumption to support a job dependency model. 

Some tricky edge cases arise, though. What happens when a task is downstream of another, but doesn't require the upstream task to succeed? E.g., a `GreatExpectations` validation job could be run on a dataset even if the upstream task fails (i.e., `TriggerRule.ALL_DONE`). Since the `OpenLineageValidationAction` fires before the upstream task's status is reported, we might assume that it read the version of the dataset produced by the upstream task, but really it's reading the older version since the upstream task failed to write. In that case, do we modify the completed GreatExpectations run to reflect the failed status of the upstream task? What if the upstream task modifies multiple datasets and one write completed successfully, but the second write failed? How do we know whether the Dataset read by GreatExpectations was the one affected by the failure? Should we assume job runs are atomic? 

Some of this is maybe outside of the scope of OpenLineage and in the domain of event consumers, but I think it's important to consider what services are doing with the data we send and offer as much clarity in the spec as possible.

**Comment by collado-mike** [comment]
> What about better supporting situations where we can tell exactly which version we're working with?

Yeah, I proposed https://github.com/OpenLineage/OpenLineage/issues/35 some months ago. This will be great for those datasets that do support explicit versioning. Unfortunately, the vast majority of data sources will not be covered by that issue.
